### ISSUE TITLE
URI Argument Injection in Dogecoin-Qt

### ISSUE THREAD

**Issue by CharesFang** [dogecoin/dogecoin#2665]
# URI Argument Injection in Dogecoin-Qt

We recently found and confirmed that Dogecoin Core in the **latest release** (version `1.14.4`)  and earlier versions are affected by a [remote code execution vulnerability in Qt5-based GUI apps](https://www.bleepingcomputer.com/news/security/qt5-based-gui-apps-susceptible-to-remote-code-execution/). Specifically, the `QApplication` in `Qt` framework used by the **Dogecoin GUI Program** will parse `Qt built-in` arguments, such as `-reverse` and `-platformpluginpath` (see the document [here](https://doc.qt.io/qt-5/qapplication.html#QApplication)). This allows a remote attacker (i.e., a web page) to invoke the local Dogecoin Core program in a victim user machine by using this kind of URI: `dogecoin:address -parameter`.

For a real attack, a malicious web page can first download a malicious Qt plugin (see [the earlier link](https://www.bleepingcomputer.com/news/security/qt5-based-gui-apps-susceptible-to-remote-code-execution/) for how to construct the plugin) to user browser's Download folder. The malicious web page then can ask user to click the URI of `dogecoin:any_wallet_address -platformpluginpath file_path_of_downloaded_plugin`. Since `dogecoin:` URI has been registered by Dogecoin in the registary, the browser will automatically launch the local Dogecoin program and invoke that downloaded plugin, causing a remote code execution.

Here for simple demonstration, we use another parameter `-reverse` to demonstrate the feasibility. By using the attached HTML page and ask user to click the URI `dogecoin:address -reverse`. It will automatically launch the local `dogecoin-qt.exe` program and you will find that the GUI has been reversed due to the parameter of `-reverse`.

We provide the screenshot and the test HTML page in the attachment.
For more details, you can refer to the following pages.

- The vulnerable code in Dogecoin: `src/qt/bitcoin.cpp`, see [line 315 to 318](https://github.com/dogecoin/dogecoin/blob/master/src/qt/bitcoin.cpp#L315-L321).
- Similar fix from Bitcoin: [PR#16578](https://github.com/bitcoin/bitcoin/pull/16578).
- The original [CVE-2021-3401](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-3401) report 
- The exploited [details](https://achow101.com/2021/02/0.18-uri-vuln).

Reported by `de957ad9679f28a38f02f00cc7928bce8fb424882ff060a3c09c32895b1474cc`.

## Attachment
1. The vulnerability test case.
![Test Case](https://user-images.githubusercontent.com/42707185/140026528-b43aa114-6ab7-42a6-ac80-c05ecb911fd6.png)

2. The test HTML page code.
```HTML
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Attack</title>
</head>
<body>

<h1>Attack in Browser</h1>
<p>Need to modify register item.</p>
<p>Here we use the -reverse parameter as a demo. For the real attack, it can use </p>
<p>
    <a href="dogecoin:3E8ociqZa9mZUSwGdSmAEMAoAxBK3FNDcd -reverse"> Click To attack </a>
</p>
</body>
</html>
```

**Comment by patricklodder** [comment]
@CharesFang Thank you! I have manually back ported the fix from Bitcoin Core and have verified locally. Could you please let me know if this fix is complete from your perspective?

**Comment by eggcn** [comment]
I hope to fix as soon as possible

**Comment by patricklodder** [comment]
> I hope to fix as soon as possible

Fix is in #2666, and we're in the last stages of release preparation so this should be out real soon.

**Comment by CharesFang** [comment]
> @CharesFang Thank you! I have manually back ported the fix from Bitcoin Core and have verified locally. Could you please let me know if this fix (#2666) is complete from your perspective?

It looks good. You can fix this vulnerability just as Bitcoin did.

**Comment by patricklodder** [comment]
#2666 is merged, will be released with 1.14.5 shortly.
